File: 백준/Gold/14497. 주난의 난（難）/주난의 난（難）.cc
#include<bits/stdc++.h>
using namespace std;
const int dy[]={-1,0,1,0};
const int dx[]={0,1,0,-1};
int n,m,y,x,yy,xx,ret;
char a[304][304];
int visited[304][304];
void bfs(int sy,int sx)
{
	visited[sy][sx]=1;
	queue<pair<int,int>> cur,next;
	cur.push({sy,sx});
	int t=1;
	while(true)
	{
		while(cur.size())
		{
			int y,x;
			tie(y,x)=cur.front();cur.pop();
			for(int i=0;i<4;i++)
			{
				int ny=y+dy[i];
				int nx=x+dx[i];
				if(ny<0 || ny>=n || nx<0 || nx>=m || visited[ny][nx]) continue;
				if(a[ny][nx]=='0' || a[ny][nx]=='#')
				{
					cur.push({ny,nx});
					visited[ny][nx]=1;
				}
				else if(a[ny][nx]=='1')
				{
					next.push({ny,nx});
					visited[ny][nx]=1;
				}
			}
		}
		if(visited[yy-1][xx-1])
		{
			cout<<t;
			return;
		}
		while(next.size())
		{
			int y,x;
			tie(y,x)=next.front();next.pop();
			cur.push({y,x});
			a[y][x]='0';
		}
		t++;
	}
}
int main()
{
	cin>>n>>m>>y>>x>>yy>>xx;
	for(int i=0;i<n;i++)
	{
		for(int j=0;j<m;j++)
		{
			cin>>a[i][j];
		}
	}
	bfs(y-1,x-1);
}

File: 백준/Gold/14497. 주난의 난（難）/주난의 난（難）.py
from collections import deque
dy=[-1,0,1,0]
dx=[0,1,0,-1]

def bfs(y,x):
    visited[y][x]=True
    cur=deque()
    nxt=deque()
    cur.append((y,x))
    ret=0
    while True:
        if visited[x2-1][y2-1]:
            print(ret)
            return

        while cur:
            sy,sx=cur.popleft()
            for i in range(4):
                ny=sy+dy[i]
                nx=sx+dx[i]
                if ny<0 or ny>=n or nx<0 or nx>=m or visited[ny][nx]:
                    continue
                if a[ny][nx] in ('0','#'):
                    visited[ny][nx]=True
                    cur.append((ny,nx))
                elif a[ny][nx]=='1':
                    visited[ny][nx]=True
                    nxt.append((ny,nx))

        while nxt:
            sy,sx=nxt.popleft()
            a[sy][sx]='0'
            cur.append((sy,sx))

        ret+=1


n,m=map(int,input().split())
x1,y1,x2,y2=map(int,input().split())

a=[[0]*m for _ in range(n)]
visited=[[False]*m for _ in range(n)]

for i in range(n):
    line=input().strip()
    for j in range(m):
        a[i][j]=line[j]

bfs(x1-1,y1-1)